### PR TITLE
add information_schema.fkinfo table to see the foreign key detail

### PR DESCRIPTION
**Pull request by juran321**
-add information_schema.fkinfo table to see the foreign key detail
-cannot show the table when table is null
-add some test case

### CODE REVIEW

File: mysql-test/output.txt [hunk 1]
```diff
@@ -0,0 +1,48 @@
+Logging: ./mysql-test-run.pl  --mem --mysqld=--rocksdb --mysqld=--default-storage-engine=rocksdb --mysqld=--skip-innodb --mysqld=--default-tmp-storage-engine=MyISAM --suite=rocksdb i_s_fkinfo.test
```
Comment (zhangquan0126) [inline mysql-test/output.txt]: Please revert the unrelated changes including this file and above files.

File: mysql-test/suite/rocksdb/r/i_s_fkinfo.result [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+CREATE TABLE t3 (t3_a INT NOT NULL, t3_b INT NOT NULL, FOREIGN KEY (t3_b) REFERENCES t2(t2_a));
+INSERT INTO t1 (t1_b) VALUES (100),(200),(300);
+INSERT INTO t2 (t2_a,t2_b,t2_c) VALUES (1,100,1),(2,200,2),(3,300,3);
+INSERT INTO t3 (t3_a, t3_b) VALUES (1,1),(2,2),(3,3);
+SELECT * FROM INFORMATION_SCHEMA.ROCKSDB_FKINFO;
+FOREIGN_TABLE_SCHEMA	FOREIGN_TABLE_NAME	FOREIGN_COLUMN_NAME	REFERENCED_TABLE_SCHEMA	REFERENCED_TABLE_NAME	REFERENCED_COLUMN_NAME	TYPE
+test	t3	t3_b	test	t2	t2_a	48
+test	t2	t2_b	test	t1	t1_b	48
+DROP TABLE t3;
+DROP TABLE t2;
```
Comment (zhangquan0126) [inline mysql-test/suite/rocksdb/r/i_s_fkinfo.result]: You can have another test to create multiple foreign keys on the same table.

File: mysql-test/var [hunk 3]
```diff
@@ -0,0 +1 @@
+/dev/shm/var_auto_JIKG
```
Comment (zhangquan0126) [inline mysql-test/var]: Same to revert.

File: storage/rocksdb/rdb_i_s.cc [hunk 4]
```diff
@@ -1042,7 +1042,163 @@ static int rdb_i_s_ddl_init(void *const p) {
...
+}// namespace RDB_FKINFO_FIELD
+
+static ST_FIELD_INFO rdb_i_s_fkinfo_fields_info[] ={
+	ROCKSDB_FIELD_INFO("FOREIGN_TABLE_SCHEMA", NAME_LEN + 1, MYSQL_TYPE_STRING, 0),
+	ROCKSDB_FIELD_INFO("FOREIGN_TABLE_NAME", NAME_LEN + 1, MYSQL_TYPE_STRING, 0),
+	ROCKSDB_FIELD_INFO("FOREIGN_COLUMN_NAME", NAME_LEN + 1, MYSQL_TYPE_STRING, 0),
+	ROCKSDB_FIELD_INFO("REFERENCED_TABLE_SCHEMA", NAME_LEN + 1, MYSQL_TYPE_STRING, 0),
+	ROCKSDB_FIELD_INFO("REFERENCED_TABLE_NAME", NAME_LEN + 1, MYSQL_TYPE_STRING, 0),
+	ROCKSDB_FIELD_INFO("REFERENCED_COLUMN_NAME", NAME_LEN + 1, MYSQL_TYPE_STRING, 0),
+	ROCKSDB_FIELD_INFO("TYPE", sizeof(uint16_t), MYSQL_TYPE_SHORT, 0),
```
Comment (zhangquan0126) [inline storage/rocksdb/rdb_i_s.cc]: better use string and convert to string for the type field as we discussed.

File: storage/rocksdb/rdb_i_s.cc [hunk 5]
```diff
@@ -1042,7 +1042,163 @@ static int rdb_i_s_ddl_init(void *const p) {
...
+	ROCKSDB_FIELD_INFO_END};
+
+//override the add_table function
+int Rdb_fkinfo_scanner::add_table(Rdb_tbl_def *tdef){
```
Comment (zhangquan0126) [inline storage/rocksdb/rdb_i_s.cc]: If we do not plan to change this pointer inside this function, and the value pointed by this pointer, better to use const Rdb_tbl_def const *tdef

File: storage/rocksdb/rdb_i_s.cc [hunk 6]
```diff
@@ -1042,7 +1042,163 @@ static int rdb_i_s_ddl_init(void *const p) {
...
+
+
+	const std::string &dbname = tdef->base_dbname();
+	const std::string &tablename = tdef->base_tablename();
+	field[RDB_FKINFO_FIELD::FOREIGN_TABLE_SCHEMA]->store(dbname.c_str(), dbname.size(),system_charset_info);
+	field[RDB_FKINFO_FIELD::FOREIGN_TABLE_NAME]->store(tablename.c_str(), tablename.size(), system_charset_info);
+	//for loop to the m_foreign_descr_set
+	const std::set<myrocks::Rdb_fk_def, myrocks::Rdb_fk_compare> &fkset = tdef->m_foreign_descr_set;
+	for(const Rdb_fk_def &fk_info : fkset){
+    Rdb_ddl_manager *ddl_manager_cur = rdb_get_ddl_manager();
```
Comment (zhangquan0126) [inline storage/rocksdb/rdb_i_s.cc]: minor: alignment.

File: storage/rocksdb/rdb_i_s.cc [hunk 7]
```diff
@@ -1042,7 +1042,163 @@ static int rdb_i_s_ddl_init(void *const p) {
...
+		GL_INDEX_ID foreign_gl_index_id = fk_info.m_foreign_gl_index_id;
+		GL_INDEX_ID referenced_gl_index_id = fk_info.m_referenced_gl_index_id;
+		//should add const in the front????
```
Comment (zhangquan0126) [inline storage/rocksdb/rdb_i_s.cc]: no need, because C++ use value copy by default.

File: storage/rocksdb/rdb_i_s.cc [hunk 8]
```diff
@@ -1042,7 +1042,163 @@ static int rdb_i_s_ddl_init(void *const p) {
...
+int Rdb_fkinfo_scanner::add_table(Rdb_tbl_def *tdef){
+	DBUG_ASSERT(tdef != nullptr);
+	
+	int ret = 0;
+
+	DBUG_ASSERT(m_table != nullptr);
+	Field **field = m_table->field;
+	DBUG_ASSERT(field != nullptr);
+
+	Rdb_dict_manager *dict_manager_cur = rdb_get_dict_manager();
```
Comment (zhangquan0126) [inline storage/rocksdb/rdb_i_s.cc]: Seems not referenced below.

File: storage/rocksdb/rdb_i_s.cc [hunk 9]
```diff
@@ -1042,7 +1042,163 @@ static int rdb_i_s_ddl_init(void *const p) {
...
+		GL_INDEX_ID foreign_gl_index_id = fk_info.m_foreign_gl_index_id;
+		GL_INDEX_ID referenced_gl_index_id = fk_info.m_referenced_gl_index_id;
+		//should add const in the front????
+		
+		std::string referenced_tablename = ddl_manager_cur->safe_get_table_name(referenced_gl_index_id);
+		std::string referenced_db;
+		std::string referenced_table;
+		std::string referenced_partname;
+    int err;
+		err = rdb_split_normalized_tablename(referenced_tablename,&referenced_db, &referenced_table, &referenced_partname);
```
Comment (zhangquan0126) [inline storage/rocksdb/rdb_i_s.cc]: just write int err =

File: storage/rocksdb/rdb_i_s.cc [hunk 10]
```diff
@@ -1042,7 +1042,163 @@ static int rdb_i_s_ddl_init(void *const p) {
...
+
+		DBUG_ASSERT(err == 0);
```
Comment (zhangquan0126) [inline storage/rocksdb/rdb_i_s.cc]: better return this failure instead of using assert. Assert will cause process crash.

File: storage/rocksdb/rdb_i_s.cc [hunk 11]
```diff
@@ -1042,7 +1042,163 @@ static int rdb_i_s_ddl_init(void *const p) {
...
+
+	DBUG_ASSERT(m_table != nullptr);
+	Field **field = m_table->field;
+	DBUG_ASSERT(field != nullptr);
+
+	Rdb_dict_manager *dict_manager_cur = rdb_get_dict_manager();
+	//Rdb_ddl_manager *ddl_manager_cur = rdb_get_ddl_manager(); // should add in the outloop or in-loop????
+
+
+	const std::string &dbname = tdef->base_dbname();
```
Comment (zhangquan0126) [inline storage/rocksdb/rdb_i_s.cc]: Seems we also need to add the database name in the information schema table?

File: storage/rocksdb/rdb_i_s.cc [hunk 12]
```diff
@@ -1780,19 +1936,19 @@ struct st_mysql_plugin rdb_i_s_ddl = {
...
-    rdb_i_s_deinit,
-    0x0001,  /* version number (0.1) */
-    nullptr, /* status variables */
-    nullptr, /* system variables */
-    nullptr, /* config options */
-    0,       /* flags */
+  MYSQL_INFORMATION_SCHEMA_PLUGIN,
+  &rdb_i_s_info,
+  "ROCKSDB_FKINFO",
+  "HAPPPPPPPPPPY",
```
Comment (zhangquan0126) [inline storage/rocksdb/rdb_i_s.cc]: You can leave your name here:)

File: storage/rocksdb/rdb_i_s.cc (hunk 9 above)
Comment (zhangquan0126) [inline storage/rocksdb/rdb_i_s.cc]: Referenced_partname is not needed here. You can check the knowledge of default parameter in C++;)

File: mysql-test/output.txt (hunk 1 above)
Comment (juran321) [reply mysql-test/output.txt]: done

File: mysql-test/suite/rocksdb/r/i_s_fkinfo.result (hunk 2 above)
Comment (juran321) [reply mysql-test/suite/rocksdb/r/i_s_fkinfo.result]: done. Add some test case for cascade

File: storage/rocksdb/rdb_i_s.cc (hunk 5 above)
Comment (juran321) [reply storage/rocksdb/rdb_i_s.cc]: this is override function, and the interface rdb_ddl_scanner::add_table is not const. 
(I tried to convert to the const, but it fail to compile :( );

File: storage/rocksdb/rdb_i_s.cc (hunk 6 above)
Comment (juran321) [reply storage/rocksdb/rdb_i_s.cc]: it shows correct on my IDE ==!

File: storage/rocksdb/rdb_i_s.cc (hunk 7 above)
Comment (juran321) [reply storage/rocksdb/rdb_i_s.cc]: done

File: storage/rocksdb/rdb_i_s.cc (hunk 9 above)
Comment (juran321) [reply storage/rocksdb/rdb_i_s.cc]: done

File: storage/rocksdb/rdb_i_s.cc (hunk 10 above)
Comment (juran321) [reply storage/rocksdb/rdb_i_s.cc]: done

File: storage/rocksdb/rdb_i_s.cc (hunk 11 above)
Comment (juran321) [reply storage/rocksdb/rdb_i_s.cc]: I think yes. cuz in InnoDB, the i_s_fk shows the database name.
It is because the foreign table and referenced table can from different database

File: storage/rocksdb/rdb_i_s.cc (hunk 12 above)
Comment (juran321) [reply storage/rocksdb/rdb_i_s.cc]: done

File: storage/rdb_i_s.cc [hunk 13]
```diff
@@ -0,0 +1,1977 @@
+/*
```
[the remark this answers is not in the document]
Comment (zhangquan0126) [reply storage/rdb_i_s.cc]: Please resolve this comment

File: storage/rocksdb/ha_rocksdb.cc [hunk 14]
```diff
@@ -7446,6 +7441,8 @@ int ha_rocksdb::create(const char *const name, TABLE *const table_arg,
           } else {
             is_on_delete = true;
           }
+        }else{
```
Comment (zhangquan0126) [inline storage/rocksdb/ha_rocksdb.cc]: minor: indent
